src/Xrefcheck/Verify.hs

```diff
@@ -327,5 +360,49 @@ checkExternalResource VerifyConfig{..} link
         HttpExceptionRequest _ exc -> case exc of
           StatusCodeException resp _
             | isAllowedErrorCode (statusCode $ responseStatus resp) -> Right ()
-            | otherwise -> Left $ ExternalResourceUnavailable (responseStatus resp)
+            | otherwise -> Left $ ExternalHttpResourceUnavailable (responseStatus resp)
           other -> Left . ExternalResourceSomeError $ show other
+
+    checkFtp :: URI -> Bool -> ExceptT VerifyError IO ()
+    checkFtp uri secure = do
+      -- get authority which stores host and port
+      authority <- case uriAuthority uri of
+        Right a -> pure a
+        Left _ -> throwError $
+          ExternalResourceInvalidUrl (Just "FTP path must be absolute")
+      let host = T.unpack . unRText $ authHost authority
+      let port :: Int = fromIntegral . fromMaybe 21 $ authPort authority
+      -- build path from pieces
+      path <- case uriPath uri of
+        Nothing -> pure "/"
+        Just (_, pieces) -> pure . T.unpack . mconcat
+          . NonEmpty.toList
+          . NonEmpty.intersperse "/"
+          . NonEmpty.map unRText
+          $ pieces
+
+      makeFtpRequest host port path secure `catch` \e ->
+        throwError $ ExternalFtpException e
+
+    makeFtpRequest
+      :: String
+      -> Int
+      -> FilePath
+      -> Bool
+      -> ExceptT VerifyError IO ()
+    makeFtpRequest host port path secure = handler host port $
+      \handle response -> do
+        -- check connection status
+        when (frStatus response /= Success) $
+          throwError $ ExternalFtpResourceUnavailable response
+        -- anonymous login
+        loginResp <- login handle "anonymous" ""
+        -- check login status
+        when (frStatus loginResp /= Success) $
+          throwError $ ExternalFtpException $ UnsuccessfulException loginResp
```

Comment:
Perhaps xrefcheck should also be able to emit warnings (not just errors) :thinking: In which case, we'd emit a warning for links requiring credentials.

And then the user could either use a flag to ignore warnings, or use a `<!-- -->` pragma to ask xrefcheck to ignore that specific link.


> So for instance, in HTTP case if I see that authentication is necessary, or a resource is forbidden for me, I still treat it as a good link because it's just that I have no access to the resource, *but some other people do*.

Nitpick: just wanted to point out that it's possible a URL might request authentication and, when authenticated, return a 404 and be an invalid link after all.

Reply:
But, to your point, I agree.

Given that we cannot emit warnings, we have a binary choice to make: treat it as valid or invalid. And we can't know for sure, so we must choose to treat it as valid to avoid false positives.

Reply:
When I think about warnings, I can't find justified usages for them in the case of xrefcheck (

If we silence them globally, then they become just redundant. If we do not silence them and allow them to be, then the project may gather so much warnings that no one ever dares to rake them out.

If we suppress a warning locally - then this warning could be just an error instead.

I dunno, looks like in Haskell we have all the same issues with warnings, and the only justified use cases for them that I yet realize are: 1. temporarily build a bad code locally; 2. a library that does not yet comply with the modern practices, e.g. some `fmt` not yet using `DerivingStrategies` extension. But in xrefcheck case, there is no sense in "temporarily bad" or "outdated" Markdown, I suppose.

Generally, CI also has no choice but to be binary, either OK or not, everything in between is an eternal backlog. And since xrefcheck is used 98% of the time in CI, not locally, producing a ternary output will probably not change much.

---

I agree that adding warnings is a very tempting option, but I just don't yet see how to implement it in a way that would work.

Maybe we actually need some different mechanism in this case :thinking: 

There was a proposal to allow `xrefcheck` analyzing only git diff. Probably here warnings would shine, but there are some fundamental difficulties in implementing that.

Reply:
I agree completely with everything you said. On second thought, there is no use-case for warnings after all.

---

That being the case, we are indeed forced to either assume an "unauthorized link" is valid or invalid, it's a binary choice.

I'm starting to think we should throw an error though... We can't prove the link is valid, so xrefcheck should fail and tell the user "401 Unauthorized - Cannot verify whether the link is valid.".

If the user is _sure_ the link is indeed valid, then they can manually add a `<!-- xrefcheck: ignore ... -->` pragma.

---

Right now, we silently ignore the 401 and simply move on. We're not even giving the user the chance to decide what to do. The link may indeed be invalid, but they'll never know.

IOW, with the current approach, we're avoiding false positives (reporting that a valid link is broken) at the expense of being able to report positives (reporting that a broken link is broken).

---

Additionally, maybe we should add a ` <!-- xrefcheck: expected <statuscode> -->` pragma? :thinking:

Reply:
> If the user is sure the link is indeed valid, then they can manually add a <!-- xrefcheck: ignore ... --> pragma to silently ignore the link

> Right now, we always silently ignore the 401 and simply move on. We're not giving the user the chance to decide what to do. The link may indeed be invalid, but they'll never know.

I think this way too. 
it's better to give possibility to decide

Reply:
You want to let the user double-check links that we cannot check?

I think it produces more hassle than a benefit, since 99% of the time I just copy the link from the browser, and now I will also have to fight with CI and add boilerplate each time :/

Maybe we could really add an option, which status codes to ignore. And maybe a global one, not as pragma? So the users can either skip 401 and 403 by default and remember to be careful, or be explicit about ignoring such links.

---

Moving to better handling true positives (finding more broken links) is good, but I think we never can get near 100% accuracy here. Another example: we cannot check anchors in external pages, that would be just too expensive. 

So to provide at least some sensible guarantees, we have to either stick to another edge (0% of false positives) or really allow configuring the behaviour here. And no way the tool should become a burden for the user unless he wants so.

Reply:
> Maybe we could really add an option, which status codes to ignore. And maybe a global one, not as pragma? So the users can either skip 401 and 403 by default and remember to be careful, or be explicit about ignoring such links.

okay, we can add only one bool config option, something like "assume protected links broken"

Reply:
done

Reply:
Nice!